crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3366,8 +3366,37 @@ impl<'db> TypeInferenceBuilder<'db> {
                     ast::CmpOp::NotIn => {
                         membership_test_comparison(MembershipTestCompareOperator::NotIn)
                     }
-                    ast::CmpOp::Is => Ok(KnownClass::Bool.to_instance(self.db)),
-                    ast::CmpOp::IsNot => Ok(KnownClass::Bool.to_instance(self.db)),
+                    ast::CmpOp::Is => {
+                        // Ideally, we would like to use `!left.is_equivalent_to(…, right)` here,
+                        // instead of `left.is_disjoint_from(…, right)`. But `is_equivalent_to` can
+                        // return false negative answers, which would lead to false `Literal[True]`
+                        // results. `is_disjoint` can also return false negatives, but we only use
+                        // positive answers for narrowing the type.
```

Comment:
The `reveal_type(a1 is o) # revealed: bool` test (and similar) make sure that we don't forget about this.

Reply:
I'm not sure the premise of this comment is quite correct? Even if `is_equivalent_to` were perfect (never returned a false negative), we could not replace `if left.is_disjoint_from(self.db, right)` with `if !left.is_equivalent_to(self.db, right)` in the code below.

If the types of `left` and `right` are disjoint, no object can be in both types, so the `is` check must be false. But if `left` and `right` are simply not-equivalent, it's still possible that they overlap, and thus we don't have any additional information about the result of the `is` check.